### PR TITLE
rewrite Friendly Date Ranges bonfire

### LINKED ISSUE #1886: Bonfire Friendly Date Ranges Test Inconsistencies
Challenge http://freecodecamp.com/challenges/bonfire-friendly-date-ranges has an issue. Please describe how to reproduce it, and include links to screenshots if possible.

I know this has been debated before, and I understand that if the dates are within 11 months it can be inferred that it is the following year, but test 4 (below) should require a year given that this is a requirement: "Secondly, if the starting year is the current year, and the ending year can be inferred by the reader, the year should be omitted."

As the 4th test is 2016/2016, it can't be inferred the year is 2016 unless there is assumed all dates are in the future, it could have been in 2015.

Could either the test condition, or the task description be changed to reflect this?

### PR DESCRIPTION
**Pull request by bheavner**
This is a pretty extensive rewrite - changed test cases, solution code, and bonfire problem statement. But I think it's internally consistent and more clear now.
closes #1886

### CODE REVIEW

**Comment by SaintPeter** [comment]
I do believe this is a significant improvement over the current definition.  Thank you so much for doing this!

I would like @FreeCodeCamp/issue-moderators to weigh in, since it is a pretty big change.

### Feedback
Please remove the function call from the bottom of solution:
`friendly(['2016-07-01', '2016-07-04']);`

The area in the instructions where you say: 
> For example, friendly(["2016-07-01", "2016-07-04"]) should return ["July 1st","4th"], and friendly(["2016-07-01", "2018-07-04"]) should return ["July 1st, 2016", "July 4th, 2018"].

The function calls and results should be wrapped in `<code>` tags:
```
For example, <code>friendly(["2016-07-01", "2016-07-04"])</code> should return <code>["July 1st","4th"]</code>, and <code>friendly(["2016-07-01", "2018-07-04"])</code> should return <code>["July 1st, 2016", "July 4th, 2018"]</code>.
```
It would be nice if this were displayed in a way that makes it easier to read, maybe as a table or by inserting `<br>` line breaks.  It's pretty hard to read as it gets all wrapped around:
image
Maybe an "Examples" section?

One last thing:  Lets make sure we test a date with 13 in it, because 13 = 13th and we need to make sure they account for that.

### Solution
Also, for reference, here is the proposed solution code:
```js
function friendly(str) {
  var thisYear = new Date().getFullYear();
  var dates = str.map(function(s) {return s.split('-').map(Number);});
  var start = dates[0];
  var end = dates[1];
  if (str[0] === str[1]) {
     return [readable(start)];
  }
  if (start[0] !== end[0]) {
    if (start[0] + 1 === end[0]){
      if (start[1] > end[1]){
        end[0] = undefined;
      }
      if (start[1] === end[1] && start[2] > end[2]){
        end[0] = undefined;
      }
      if (start[0] === thisYear){
        start[0] = undefined;
      }
    }
    return dates.map(readable);
  }
  end[0] = undefined;
  if (start[0] === thisYear){
    start[0] = undefined;
  }
  if (start[1] === end[1]) {
    end[1] = undefined;
  }
  return dates.map(readable);
}

function readable(arr) {
  var ordD = arr[2] + nth(arr[2]);
  if (!arr[1]) {
    return ordD;
  }
  return MONTH[arr[1]] + " " + ordD + (!arr[0] ? "" : ", " + arr[0]);
}

var MONTH = {1: "January",
             2: "February",
             3: "March",
             4: "April",
             5: "May",
             6: "June",
             7: "July",
             8: "August",
             9: "September",
             10: "October",
             11: "November",
             12: "December"};

function nth(d) {
  if(d>3 && d<21) return 'th';
  switch (d % 10) {
    case 1: return "st";
    case 2: return "nd";
    case 3: return "rd";
    default: return "th";
  }
}
```

**Comment by bheavner** [comment]
I think I've address your suggestions, though I didn't dig too deep into formatting of the examples - just added the `<code>` and line breaks. I changed the January test to use the 13th instead of the 1st.

**Comment by SaintPeter** [comment]
Changes look solid.  Just waiting for feedback from the other mods.

File: seed/challenges/01-front-end-development-certification/advanced-bonfires.json [hunk 1]
```diff
@@ -320,30 +286,25 @@
...
...){\n      if (start[1] > end[1]){\n        end[0] = undefined;\n      }\n      if (start[1] === end[1] && start[2] > end[2]){\n        end[0] = undefined;\n      }\n      if (start[0] === thisYear){\n        start[0] = undefined;\n      }\n    }\n    return dates.map(readable);\n  }\n  end[0] = undefined;\n  if (start[0] === thisYear){\n    start[0] = undefined;\n  }\n  if (start[1] === end[1]) {\n    end[1] = undefined;\n  }\n  return dates.map(readable);\n}\n\nfunction readable(arr) {\n  var ordD = arr[2] + nth(arr[2]);\n  if (!arr[1]) {\n    return ordD;\n  }\n  return MONTH[arr[1]] + \" \" + ordD + (!arr[0] ? \"\" : \", \" + arr[0]);\n}\n\nvar MONTH = {1: \"January\",\n             2: \"February\",\n             3: \"March\",\n             4: \"April\",\n             5: \"May\",\n             6: \"June\",\n             7: \"July\",\n             8: \"August\",\n             9: \"September\",\n             10: \"October\",\n             11: \"November\",\n             12: \"December\"};\n\nfunction nth(d) {\n  if(d>3 && d<21) return 'th';\n  switch (d % 10) {\n    case 1: return \"st\";\n    case 2: return \"nd\";\n    case 3: return \"rd\";\n    default: return \"th\";\n  }\n}"
```
Comment (bugron) [inline seed/challenges/01-front-end-development-certification/advanced-bonfires.json]: This solution isn't passing all tests. @bheavner, @SaintPeter can you correct it, please?

Comment (SaintPeter) [inline seed/challenges/01-front-end-development-certification/advanced-bonfires.json]: I just ran this manually as well as with `npm run test-challenges` and it worked fine for me.

**Comment by bugron** [comment]
@bheavner thanks, merged.

File: seed/challenges/01-front-end-development-certification/advanced-bonfires.json (hunk 1 above)
Comment (bugron) [inline seed/challenges/01-front-end-development-certification/advanced-bonfires.json]: That was my fault. Sorry.

**Pull request merged**